File: lib/contracts/content.ts
/**
 * Content Engine contracts (platform doc §4.1, §4.4).
 *
 * The source doc locks the ContentProvider Protocol method shapes
 * (get_course, get_playback_manifest) but NOT the full Course /
 * SignedManifest / Enrollment field lists — those come from the `courses`,
 * `lessons`, `enrollments` table references. Field choices here are
 * reasonable decisions, logged in the assumption register (index.ts).
 */

export type CourseLevel = "beginner" | "intermediate" | "advanced";
export type LessonKind = "video" | "article";
export type CourseFormat = "video" | "interactive" | "lab" | "project" | "judge";
export type CareerTrack = "cyber_security" | "web_development" | "ai_ml" | "cloud" | "data_science" | "game_dev" | "interview_prep";
export type DurationFilter = "under_2" | "2_to_5" | "5_to_10" | "over_10";
export type CourseSort = "popular" | "rated" | "newest" | "recommended" | "shortest";
/**
 * §4.4 — courses are versioned like code: draft vs published.
 * F7 (Admin/CMS) adds `in_review` for the submit-for-review →
 * second-reviewer-publish workflow (build.md F7).
 */
export type ContentStatus = "draft" | "in_review" | "published";
export type EnrollmentStatus = "active" | "completed";

export interface CourseInstructor {
  id: string;
  display_name: string;
  title: string;
}

export interface CourseLesson {
  id: string;
  title: string;
  kind: LessonKind;
  /** Seconds of video; 0 for articles */
  duration_seconds: number;
  position: number;
  /** Guest-accessible lesson preview, served by the Content Engine. */
  isPreview: boolean;
  /** Mock stand-in for the authored lesson body. */
  preview_body?: string | null;
}

export interface LessonPreview {
  lesson_id: string;
  title: string;
  kind: LessonKind;
  duration_seconds: number;
  body: string;
  manifest_url: string | null;
}

/** Content Engine review projection (build.md F1 course reviews). */
export interface Review {
  id: string;
  course_id: string;
  author: {
    id: string;
    name: string;
    avatar_url: string | null;
  };
  rating: number;
  date: string;
  comment: string;
  helpful_count: number;
}

export interface ReviewPage {
  course_id: string;
  offset: number;
  total: number;
  reviews: Review[];
  has_more: boolean;
}

export interface CourseSection {
  id: string;
  title: string;
  position: number;
  lessons: CourseLesson[];
}

export interface Course {
  id: string;
  title: string;
  subtitle: string;
  description: string;
  category: string;
  level: CourseLevel;
  language: string;
  /** §4.4 draft vs published */
  status: ContentStatus;
  /**
   * F7 review workflow — who submitted the current revision for review.
   * Drives the two-person rule (publisher must differ from the submitter).
   * Null until a draft is submitted. Provisional, logged in the register.
   */
  submitted_by?: string | null;
  /** F7 review workflow — who published the last version (second reviewer). */
  reviewed_by?: string | null;
  instructor: CourseInstructor;
  rating: number;
  review_count: number;
  /** 0 = free */
  price_cents: number;
  enrolled_count: number;
  estimated_hours: number;
  syllabus: CourseSection[];
  created_at: string;
  updated_at: string;
  format?: CourseFormat;
  career_track?: CareerTrack;
  is_project_based?: boolean;
  certificate_included?: boolean;
}

/** Lightweight card shape for catalog hits */
export interface CourseSummary {
  id: string;
  title: string;
  subtitle: string;
  category: string;
  level: CourseLevel;
  rating: number;
  review_count: number;
  price_cents: number;
  enrolled_count: number;
  estimated_hours: number;
  instructor_name: string;
  language: string;
  /** Seed for the gradient cover art (mock stand-in for poster art) */
  cover_hue: number;
  format: CourseFormat;
  career_track: CareerTrack;
  is_project_based: boolean;
  certificate_included: boolean;
}

/**
 * SignedManifest — signed, short-TTL HLS URL (platform §2.3/§4.1).
 * In mock mode `signature` is a deterministic stand-in; the real backend
 * signs with a server-held key and enforces expiry (a fetch after
 * `expires_at` must 403).
 */
export interface SignedManifest {
  lesson_id: string;
  user_id: string;
  manifest_url: string;
  expires_at: string;
  signature: string;
  /** WebVTT captions URL, null when the lesson has none */
  captions_url: string | null;
}

export interface Enrollment {
  course_id: string;
  user_id: string;
  status: EnrollmentStatus;
  /** 0-100 across all lessons */
  progress_pct: number;
  /** Resume position */
  last_lesson_id: string | null;
  last_position_seconds: number;
  enrolled_at: string;
  updated_at: string;
}

/* ------------------------------------------------------------------ */
/*  Catalog search — mock Meilisearch response shape                    */
/*  (mirrors the real Meilisearch JSON API field names so the later     */
/*  swap to the self-hosted Meilisearch instance is field-identical)    */
/* ------------------------------------------------------------------ */

export interface CatalogFacet {
  category: string;
  count: number;
}

export interface MeilisearchCatalogResponse {
  hits: CourseSummary[];
  query: string;
  processingTimeMs: number;
  limit: number;
  offset: number;
  estimatedTotalHits: number;
}

export interface CatalogQuery {
  query?: string;
  category?: string;
  price?: "free" | "paid" | "all";
  level?: CourseLevel | "all";
  page?: number;
  pageSize?: number;
  duration?: DurationFilter;
  format?: CourseFormat | "all";
  careerTrack?: CareerTrack | "all";
  projectBased?: boolean;
  certificateIncluded?: boolean;
  minRating?: number;
  sort?: CourseSort;
}

File: backend/platform_core/contracts/content.py
"""Content Engine contracts (platform §4.1, §4.4) — Pydantic mirror of `lib/contracts/content.ts`,
field-for-field. The `ContentProvider` Protocol method shapes are locked by the doc; full
`Course`/`SignedManifest`/`Enrollment` field lists are the frontend's own logged, reasonable
decisions, reproduced here verbatim so backend and frontend never silently diverge.
"""

from datetime import datetime
from typing import Literal, Protocol
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

CourseLevel = Literal["beginner", "intermediate", "advanced"]
LessonKind = Literal["video", "article"]
CourseFormat = Literal["video", "interactive", "lab", "project", "judge"]
CareerTrack = Literal[
    "cyber_security",
    "web_development",
    "ai_ml",
    "cloud",
    "data_science",
    "game_dev",
    "interview_prep",
]
ContentStatus = Literal["draft", "in_review", "published"]
EnrollmentStatus = Literal["active", "completed"]


class CourseInstructor(BaseModel):
    id: UUID
    display_name: str
    title: str


class CourseLesson(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: UUID
    title: str
    kind: LessonKind
    duration_seconds: int
    position: int
    # `lib/contracts/content.ts` spells this camelCase (the one exception to that file's
    # otherwise-uniform snake_case) — kept as an alias so wire compatibility with the frontend
    # doesn't require the whole codebase to violate PEP 8 naming.
    is_preview: bool = Field(alias="isPreview")
    preview_body: str | None = None


class CourseSection(BaseModel):
    id: UUID
    title: str
    position: int
    lessons: list[CourseLesson]


class Course(BaseModel):
    id: UUID
    title: str
    subtitle: str
    description: str
    category: str
    level: CourseLevel
    language: str
    status: ContentStatus
    submitted_by: UUID | None = None
    reviewed_by: UUID | None = None
    instructor: CourseInstructor
    rating: float
    review_count: int
    price_cents: int
    enrolled_count: int
    estimated_hours: float
    syllabus: list[CourseSection]
    created_at: datetime
    updated_at: datetime
    format: CourseFormat | None = None
    career_track: CareerTrack | None = None
    is_project_based: bool | None = None
    certificate_included: bool | None = None


class CourseSummary(BaseModel):
    """Lightweight card shape for catalog hits."""

    id: UUID
    title: str
    subtitle: str
    category: str
    level: CourseLevel
    rating: float
    review_count: int
    price_cents: int
    enrolled_count: int
    estimated_hours: float
    total_lessons: int
    instructor_name: str
    language: str
    format: CourseFormat
    career_track: CareerTrack
    is_project_based: bool
    certificate_included: bool


class SignedManifest(BaseModel):
    """Signed, short-TTL HLS URL (platform §2.3/§4.1). A fetch after `expires_at` must 403."""

    lesson_id: UUID
    user_id: UUID
    manifest_url: str
    expires_at: datetime
    signature: str
    captions_url: str | None


class Enrollment(BaseModel):
    course_id: UUID
    user_id: UUID
    status: EnrollmentStatus
    progress_pct: float
    last_lesson_id: UUID | None
    last_position_seconds: int
    enrolled_at: datetime
    updated_at: datetime


class MeilisearchCatalogResponse(BaseModel):
    """Mirrors the real Meilisearch JSON API field names verbatim."""

    hits: list[CourseSummary]
    query: str
    processingTimeMs: int  # noqa: N815 - verbatim Meilisearch field name
    limit: int
    offset: int
    estimatedTotalHits: int  # noqa: N815 - verbatim Meilisearch field name


class ContentProvider(Protocol):
    """Platform §4.1 — locked verbatim."""

    async def get_course(self, course_id: UUID) -> Course: ...

    async def get_playback_manifest(self, lesson_id: UUID, user_id: UUID) -> SignedManifest: ...
